pingen/pingen_document.py

```diff
@@ -141,186 +135,168 @@ def _push_to_pingen(self, cr, uid, document, pingen=None, context=None):
-
         Commit after each record
-
         Instead of raising, store the error in the pingen.document
         """
-        if not ids:
-            ids = self.search(
-                cr, uid,
-                # do not retry pingen_error, they should be treated manually
-                [('state', 'in', ['pending', 'pushed', 'error'])],
-                limit=100,
-                context=context)
-
-        with closing(pooler.get_db(cr.dbname).cursor()) as loc_cr, \
-                self._get_pingen_session(cr, uid, context=context) as session:
-            for document in self.browse(loc_cr, uid, ids, context=context):
-
-                if document.state == 'error':
-                    self._resolve_error(loc_cr, uid, document, context=context)
-                    document.refresh()
-
-                try:
-                    if document.state == 'pending':
-                        self._push_to_pingen(
-                            loc_cr, uid, document, pingen=session, context=context)
-
-                    elif document.state == 'pushed':
-                        self._ask_pingen_send(
-                            loc_cr, uid, document, pingen=session, context=context)
-                except ConnectionError as e:
-                    document.write({'last_error_message': e,
-                                    'state': 'error'},
-                                   context=context)
-                except APIError as e:
-                    document.write({'last_error_message': e,
-                                    'state': 'pingen_error'},
-                                   context=context)
-                except:
-                    _logger.error('Unexcepted error in pingen cron')
-                    loc_cr.rollback()
-                    raise
-
-                else:
-                    loc_cr.commit()
-
+        cr = registry(self._cr.dbname).cursor()
+        self = self.with_env(self.env(cr=cr))
+        for document in self:
+            if document.state == 'error':
+                self._resolve_error(document)
+                document.refresh()
+            try:
+                if document.state == 'pending':
+                    self._push_to_pingen(document)
+
+                elif document.state == 'pushed':
+                    self._ask_pingen_send(document)
+            except ConnectionError as e:
+                document.write({'last_error_message': e,
+                                'state': 'error'})
+            except APIError as e:
+                document.write({'last_error_message': e,
+                                'state': 'pingen_error'})
+            except:
+                _logger.error('Unexcepted error in pingen cron')
+                self.env.cr.rollback()
+                raise
+            else:
+                self.env.cr.commit()
         return True
 
-    def _resolve_error(self, cr, uid, document, context=None):
+    def _resolve_error(self, document):
```

Comment:
For all method of `ir.attachment` having `document` as parameter you can remove `document` from the method signature and use `self` instead.

Then you will be able to call them in a more object oriented maneer:

document._resolve_error()